### PR TITLE
Better Spells Vol. 2

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Improves the QOL of all targetted spells.
 - Targetted spells that require you to select a target will no longer make you pick from a list, but will give you a ranged ability so you can click on the target you want to affect.

Improves the cooldown overlay
 - Cooldown overlay maptext and the spinning bar thing is now on top of the icon.

## Why It's Good For The Game

Quality of life for spells, much easier to use.

## Changelog
:cl:
tweak: Targeted spells now give you a click ability rather than a dropdown list of targets.
tweak: The cooldown on spells is now on top of the button.
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
Need to perfect it a little by shifting some things and fixing the timer animation on spells with 0 max_charge

File: code/modules/spells/spell_types/barnyard.dm [hunk 1]
```diff
@@ -4,7 +4,6 @@
 	school = "transmutation"
 	charge_type = "recharge"
 	charge_max	= 150
-	charge_counter = 0
```
Comment (PowerfulBacon) [inline code/modules/spells/spell_types/barnyard.dm:7]: pointless code, doesn't do anything because the charge is set to max in init

**Comment by PowerfulBacon** [comment]
All of the spells that manually add targeting like mindswap probably could just be thrown into this, but I'm lazy

**Comment by eeSPee** [comment]
Can you also port the default spell/cone from the TG code in this PR, please?

**Comment by PowerfulBacon** [comment]
> 
> 
> Can you also port the default spell/cone from the TG code in this PR, please?

What is that?
Also I really hate porting and would prefer just to code it myself.

**Comment by eeSPee** [comment]
> > Can you also port the default spell/cone from the TG code in this PR, please?
> 
> What is that?
> Also I really hate porting and would prefer just to code it myself.

A spell that does what affects all turfs in an arc from the caster. It's used in some fire breath spells I think, and the new heretic rust spell.

**Comment by park66665** [comment]
This PR touches how spells are recharged, so these codes might need your touch:

Golem spells you probably missed:

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/mob/living/carbon/human/species_types/golems.dm#L1069-L1077

`disrupt_spell` might won't trigger recharge as needed; `delay` often is lower than 0:

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/datums/mind.dm#L663-L671

`/obj/effect/proc_holder/spell/targeted/touch` code in general.

Insta-recharging codes that probably won't need your attention:

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/projectiles/projectile/magic.dm#L543-L546

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/datums/status_effects/neutral.dm#L114-L117

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/spells/spell_types/soultap.dm#L30-L33

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/game/objects/structures/ghost_role_spawners.dm#L390-L404

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/antagonists/revenant/revenant_abilities.dm#L159-L171

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/antagonists/wizard/equipment/spellbook.dm#L40-L44

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/events/wizard/aid.dm#L38-L45

https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/mob/living/simple_animal/constructs.dm#L225-L226

Otherwise PR looks okay to me.

**Comment by PowerfulBacon** [comment]
> 
> 
> This PR touches how spells are recharged, so these codes might need your touch:
> 
> Golem spells you probably missed:
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/mob/living/carbon/human/species_types/golems.dm#L1069-L1077
> 
> `disrupt_spell` might won't trigger recharge as needed; `delay` often is lower than 0:
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/datums/mind.dm#L663-L671
> 
> `/obj/effect/proc_holder/spell/targeted/touch` code in general.
> 
> Insta-recharging codes that probably won't need your attention:
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/projectiles/projectile/magic.dm#L543-L546
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/datums/status_effects/neutral.dm#L114-L117
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/spells/spell_types/soultap.dm#L30-L33
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/game/objects/structures/ghost_role_spawners.dm#L390-L404
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/antagonists/revenant/revenant_abilities.dm#L159-L171
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/antagonists/wizard/equipment/spellbook.dm#L40-L44
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/events/wizard/aid.dm#L38-L45
> 
> https://github.com/BeeStation/BeeStation-Hornet/blob/9a38b3286debedbb8e30ef73cabb0010752cb9f3/code/modules/mob/living/simple_animal/constructs.dm#L225-L226
> 
> Otherwise PR looks okay to me.

These ones should be ok. The ones I removed was the charge_counter = 0 being in the define for the subtype since on New() it gets set to max anyway making the code pointless (unless I'm missing something)

**Comment by PowerfulBacon** [comment]
oh but yea the other ones are affected

**Comment by PowerfulBacon** [comment]
Seems to be working good now

**Review by park66665 — CHANGES_REQUESTED** [review]
After some testing, it was revealed that `/obj/effect/proc_holder/spell/targeted/touch` spells don't work properly with this PR:

1. They instantly recharge when the hand is used
2. They need recharging when the hand is retrieved

**Comment by PowerfulBacon** [comment]
Alright working on improving the touch spells

File: code/modules/spells/spell_types/touch_attacks.dm [hunk 2]
```diff
@@ -6,15 +6,17 @@
...
 /obj/effect/proc_holder/spell/targeted/touch/Destroy()
 	remove_hand()
 	to_chat(usr, "<span class='notice'>The power of the spell dissipates from your hand.</span>")
 	..()
 
-/obj/effect/proc_holder/spell/targeted/touch/proc/remove_hand(recharge = FALSE)
+/obj/effect/proc_holder/spell/targeted/touch/proc/remove_hand(mob/user)
 	QDEL_NULL(attached_hand)
-	if(recharge)
+	if(!spell_used)
```
Comment (park66665) [inline code/modules/spells/spell_types/touch_attacks.dm]: `user` argument is not used.

File: code/modules/spells/spell_types/godhand.dm [hunk 3]
```diff
@@ -186,8 +187,4 @@
 		*/
 		M.forceMove(pie)
 
-
-		charges--
-
-	if(charges <= 0)
-		qdel(src)
+	. = ..()
```
Comment (park66665) [inline code/modules/spells/spell_types/godhand.dm]: You need to clean `/obj/item/melee/touch_attack/bspie/afterattack`. It still has playsound and the parent proc call position doesn't seems to be okay.

File: code/modules/admin/verbs/debug.dm [hunk 4]
```diff
@@ -900,3 +900,11 @@ But you can call procs that are of type /mob/living/carbon/human/proc/ for that
 	message_admins("<span class='adminnotice'>[key_name_admin(src)] modified \the [C.name] at [AREACOORD(C)] - Gas: [gas_to_add], Moles: [amount], Temp: [temp].</span>")
 	log_admin("[key_name_admin(src)] modified \the [C.name] at [AREACOORD(C)] - Gas: [gas_to_add], Moles: [amount], Temp: [temp].")
 
+/client/proc/give_all_spells()
+	set category = "Debug"
+	set name = "Give all spells"
+	if(!check_rights(R_DEBUG))
+		return
+	for(var/type in subtypesof(/obj/effect/proc_holder/spell))
```
Comment (park66665) [inline code/modules/admin/verbs/debug.dm]: There is a global called `GLOB.spells` that is a similar list iirc.

**Review by park66665 — COMMENTED** [review]
Things look... okay-ish, I guess?

File: code/modules/spells/spell_types/godhand.dm (hunk 3 above)
Comment (PowerfulBacon) [reply code/modules/spells/spell_types/godhand.dm]: will move parent inside the do_mob

File: code/modules/spells/spell_types/touch_attacks.dm (hunk 2 above)
Comment (PowerfulBacon) [reply code/modules/spells/spell_types/touch_attacks.dm]: resolved

File: code/modules/admin/verbs/debug.dm [hunk 5]
```diff
@@ -905,6 +905,6 @@ But you can call procs that are of type /mob/living/carbon/human/proc/ for that
 	set name = "Give all spells"
 	if(!check_rights(R_DEBUG))
 		return
-	for(var/type in subtypesof(/obj/effect/proc_holder/spell))
-		var/obj/effect/proc_holder/spell/spell = new type
+	for(var/type in GLOB.spells)
+		var/obj/effect/proc_holder/spell/spell = type
```
Comment (park66665) [inline code/modules/admin/verbs/debug.dm]: I am afraid you'll still need `new`

Comment (PowerfulBacon) [reply code/modules/admin/verbs/debug.dm]: Is glob.spells not already initialized?

Comment (PowerfulBacon) [reply code/modules/admin/verbs/debug.dm]: fuck its not

**Review by park66665 — APPROVED** [review]
All clear. This PR is good to go, finally.

**Pull request merged**